### PR TITLE
Fetch ignition config from S3

### PR DESCRIPTION
**Pull request by beyondbill**
Fixes https://takescoop.atlassian.net/browse/SEC-8

Instead of passing ignition config (which contains secrets) into EC2 instances via userdata (which is available within every k8s pod via EC2 metadata endpoint), we now pass in a ignition config as userdata which will download the actual ignition config from S3.

### CODE REVIEW

**Comment by beyondbill** [comment]
@bendrucker Any idea on how to test this? I can hook the commit in kubernetes repo and apply in playground. But I'd like to know whether there's a better way.

File: aws/container-linux/kubernetes/controllers.tf [hunk 1]
```diff
@@ -89,20 +108,34 @@ data "template_file" "etcds" {
...
+  name   = "read-ignition-configs"
+  role   = "${aws_iam_role.controller.id}"
+  policy = "${data.aws_iam_policy_document.controller_read_ignition_configs.json}"
+}
+
+data "aws_iam_policy_document" "controller_read_ignition_configs" {
+  statement {
+    effect    = "Allow"
+    actions   = ["s3:GetObject"]
+    resources = "${formatlist("${aws_s3_bucket.ignition_configs.arn}/%v", aws_s3_bucket_object.controller-ignitions.*.id)}"
```
Comment (beyondbill) [inline aws/container-linux/kubernetes/controllers.tf]: This does not work.

```
Error: module.lego.module.typhoon.data.aws_iam_policy_document.controller_read_ignition_configs: statement.0.resources: should be a list 
```

**Comment by beyondbill** [comment]
@bendrucker I managed to successfully run `terraform plan` on the existing playground cluster, which means that all syntax issues have been fixed. Will create a new cluster for deeper testing tomorrow.

**Comment by beyondbill** [comment]
We decide to postpone this task a sprint to focus on something else (e.g. SOC2). For my own record, testing to create a new cluster in playground failed

```
/app/environments/playground # terraform apply --target module.test-typhoon
````

```
......
Error: Error applying plan:

5 errors occurred:
	* module.test-typhoon.module.typhoon.module.workers.aws_iam_role_policy.worker_read_ignition_configs: 1 error occurred:
	* aws_iam_role_policy.worker_read_ignition_configs: Error putting IAM role policy read-ignition-configs: MalformedPolicyDocument: Resource test-typhoon-ignition-configs/worker.yaml must be in ARN format or "*".
	status code: 400, request id: eb9eed93-be6d-4cd5-a79f-011da3e17891


	* module.test-typhoon.module.peering_main.aws_route.accepter_to_requester[2]: 1 error occurred:
	* aws_route.accepter_to_requester.2: Error creating route: RouteAlreadyExists: The route identified by 10.10.16.0/20 already exists.
	status code: 400, request id: dd6d4c16-3106-4266-8c1c-30d62267fe53


	* module.test-typhoon.module.peering_main.aws_route.accepter_to_requester[1]: 1 error occurred:
	* aws_route.accepter_to_requester.1: Error creating route: RouteAlreadyExists: The route identified by 10.10.16.0/20 already exists.
	status code: 400, request id: e8dcaeb4-f84e-413d-9ff2-8d877c172e21


	* module.test-typhoon.module.typhoon.aws_autoscaling_group.bastion: 1 error occurred:
	* aws_autoscaling_group.bastion: "test-typhoon-bastion terraform-20200114173735200800000007": Waiting up to 10m0s: Need at least 1 healthy instances in ELB, have 0. Most recent activity: {
  ActivityId: "79e5be94-89fc-a741-5cff-b9f919513637",
  AutoScalingGroupName: "test-typhoon-bastion terraform-20200114173735200800000007",
  Cause: "At 2020-01-14T17:38:17Z an instance was started in response to a difference between desired and actual capacity, increasing the capacity from 0 to 1.",
  Description: "Launching a new EC2 instance: i-0eeb4f77e048214dc",
  Details: "{\"Subnet ID\":\"subnet-080e0159b302a74d0\",\"Availability Zone\":\"us-west-2c\"}",
  EndTime: 2020-01-14 17:38:51 +0000 UTC,
  Progress: 100,
  StartTime: 2020-01-14 17:38:19.049 +0000 UTC,
  StatusCode: "Successful"
}


	* module.test-typhoon.module.peering_main.aws_route.accepter_to_requester[0]: 1 error occurred:
	* aws_route.accepter_to_requester.0: Error creating route: RouteAlreadyExists: The route identified by 10.10.16.0/20 already exists.
	status code: 400, request id: 90589754-58c2-4806-819a-998436a278bb





Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.



Error: failed to retrieve lock info: unexpected end of JSON input
```

File: aws/container-linux/kubernetes/bastion.tf [hunk 2]
```diff
@@ -1,3 +1,16 @@
+locals {
+  user_data = {
```
Comment (bendrucker) [inline aws/container-linux/kubernetes/bastion.tf]: Locals are module-scoped and you're setting `user_data` multiple times. I think you should be able to do this inline.

File: aws/container-linux/kubernetes/controllers.tf [hunk 3]
```diff
@@ -1,3 +1,16 @@
+locals {
+  user_data = {
+    ignition = {
+      version = "2.3.0"
+      config = {
+        replace = {
+          source = "s3://${aws_s3_bucket.ignition_configs.id}/${element(aws_s3_bucket_object.controller-ignitions.*.id, count.index)}"
```
Comment (bendrucker) [inline aws/container-linux/kubernetes/controllers.tf]: You can't use `count.index` in this context, this must be inside `aws_instance.controllers`. If passing a map inline to `jsonencode` doesn't work then I'd just revert to the JSON string you had before.

File: aws/container-linux/kubernetes/controllers.tf [hunk 4]
```diff
@@ -89,20 +108,34 @@ data "template_file" "etcds" {
...
+  name   = "read-ignition-configs"
+  role   = "${aws_iam_role.controller.id}"
+  policy = "${data.aws_iam_policy_document.controller_read_ignition_configs.json}"
+}
+
+data "aws_iam_policy_document" "controller_read_ignition_configs" {
+  statement {
+    effect    = "Allow"
+    actions   = ["s3:GetObject"]
+    resources = ["${formatlist("${aws_s3_bucket.ignition_configs.arn}/%v", aws_s3_bucket_object.controller-ignitions.*.id)}"]
```
Comment (bendrucker) [inline aws/container-linux/kubernetes/controllers.tf]:
```suggestion
    resources = ["${formatlist("${aws_s3_bucket.ignition_configs.arn}/%s", aws_s3_bucket_object.controller-ignitions.*.id)}"]
```

**Comment by beyondbill** [comment]
@bendrucker I successfully created a playground cluster, was able to connect via bastion and to list pods. We can sign this off and wait for the next cluster migration!

**Comment by bendrucker** [comment]
@beyondbill If you resolve the conflicts here, happy to try to get this released!

**Comment by beyondbill** [comment]
@bendrucker All conflicts resolved!

**Pull request merged**